README.md
#validations

The best project ever.


CHANGELOG.md

0.0.6 / 2015-01-27
==================

  * Merge pull request #8 from firstandthird/feature/additional-validations
  * Added max length and url validations

0.0.5 / 2015-01-22
==================

  * Merge pull request #6 from firstandthird/bug/required
  * Trims text before checking length
  * Changed required validation to accept more types.
  * Check if input is a string before checking length.
  * Add tests for slug
  * Add slug validation. Limit to lowercase, dash, numbers.

0.0.4 / 2014-03-26 
==================

  * added support for capital letters in email address

0.0.3 / 2014-01-07 
==================

  * chore(dist): rebuilt
  * feat(bower): more meta data
  * feat(grunt): don't generate full files

0.0.2 / 2014-01-03
==================

 * I'm stupid and changed the dist file :poop:

0.0.1 / 2014-01-03
==================

 * First commit.


test/validations.test.js
var validator = new validations({
  urlCustom: function(input) {
    return validations.validate(input, function(text) {
      return text.indexOf('htt') === 0;
    }, 'urlCustom');
  }
});

suite('validations', function() {
  test('required', function(){
    assert.ok(validations.required('test').valid);
    assert.ok(validations.required(['something']).valid);
    assert.ok(validations.required({something: 'value'}).valid);
    assert.ok(validations.required(1234).valid);
    assert.ok(validations.required(true).valid);
    assert.ok(validations.required(false).valid);
    assert.ok(validations.required(' test ').valid);
    assert.ok(!validations.required('').valid);
    assert.ok(!validations.required(null).valid);
    assert.ok(!validations.required().valid);
    assert.ok(!validations.required([]).valid);
    assert.ok(!validations.required().valid);
    assert.ok(!validations.required(' ').valid);
  });

  test('slug', function(){
    assert.ok(validations.slug('my-slug-2014').valid);
    assert.ok(!validations.slug('My-Slug-2014').valid);
    assert.ok(!validations.slug('my_slug-2014').valid);
    assert.ok(!validations.slug('my slug 2014').valid);
    assert.ok(!validations.slug('my-slug-2014!').valid);
  });

  test('email', function(){
    assert.ok(validations.email('email@example.com').valid);
    assert.ok(validations.email('Email@Example.com').valid);
    assert.ok(!validations.email('example.com').valid);
  });

  test('password', function(){
    assert.ok(validations.password('V3ZjuXLLxnUR9y').valid);
    assert.ok(!validations.password('pass').valid);
  });

  test('passwordConfirm', function(){
    assert.ok(validations.passwordConfirm('V3ZjuXLLxnUR9y', 'V3ZjuXLLxnUR9y').valid);
    assert.ok(!validations.passwordConfirm('pass', 'V3ZjuXLLxnUR9y').valid);
  });

  test('passwordStrength', function(){
    assert.ok(validations.passwordStrength('V3ZjuXLLxnUR9y') > 99);
    assert.ok(validations.passwordStrength('iEad2mc9E') > 50);
    assert.ok(validations.passwordStrength('pass') < 50);
  });

  test('max length', function(){
    assert.ok(validations.maxLength('test', 5).valid);
    assert.ok(!validations.maxLength('test', 2).valid);
  });
  
  test('url', function(){
    assert.ok(validations.url('http://google.com').valid);
    assert.ok(validations.url('https://google.com').valid);
    assert.ok(validations.url('https://google.com?query=true').valid);
    assert.ok(!validations.url('google.com').valid);
  });

  test('custom: urlCustom', function(){
    assert.ok(validations.urlCustom('http://google.com').valid);
    assert.ok(validations.urlCustom('https://google.com').valid);
    assert.ok(!validations.urlCustom('google.com').valid);
  });
});


lib/validations.js
(function(){
  var root = this;

  var alphaSize = function(text) {
    var count = 0;
    var lower, upper, number, symbol, symbol2;
    lower = upper = number = symbol = symbol2 = false;
    var other = '';
    var i, c;

    for(i = 0, c = text.length; i < c; i++) {
      if(!lower && /[a-z]/.test(text)) {
        count += 26;
        lower = true;
      } else if(!upper && /[A-Z]/.test(text)) {
        count += 26;
        upper = true;
      } else if(!number && /\d/.test(text)) {
        count += 10;
        number = true;
      } else if(!symbol && '!@#$%^&*()'.indexOf(text) > -1) {
        count += 10;
        symbol = true;
      } else if(!symbol2 && '~`-_=+[]{}\\|;:\'",.<>?/'.indexOf(text) > -1) {
        count += 22;
        symbol2 = true;
      } else if(other.indexOf(text) === -1) {
        count++;
        other += text;
      }
    }

    return count;
  };

  var trimString = function(text) {
    return text.replace(/^[\s\uFEFF\xA0]+|[\s\uFEFF\xA0]+$/g, '');
  };

  var validations = function(customValidators) {
    var validator;

    if(typeof customValidators === 'object') {
      for(validator in customValidators) {
        if(!validations[validator]) {
          validations[validator] = customValidators[validator];
        }
      }
    }

    return this;
  };

  validations.validate = function(input, pattern, type) {
    var result = {
      valid: false,
      type: type || 'validate'
    };

    if(pattern instanceof RegExp) {
      result.valid = pattern.test(input);
    } else if(typeof pattern === 'string') {
      result.valid = pattern === input;
    } else if(typeof pattern === 'function') {
      result.valid = pattern(input);
    }

    return result;
  };

  validations.required = function(input) {
    return validations.validate(input, function(text) {
      var type = typeof text;

      if(type === 'undefined' || text === null) {
        return false;
      }

      if(text instanceof Array) {
        return text.length > 0;
      }

      if(type === 'string') {
        return trimString(text).length > 0;
      }

      if(type === 'object') {
        
        for(var i in text) {
          if(text.hasOwnProperty(i)) {
            return true;
          }
        }

        return false;
      }

      return true;
    }, 'required');
  };

  validations.slug = function(input) {
    return validations.validate(input, /^[a-z0-9-]+$/, 'slug');
  };

  validations.email = function(input) {
    return validations.validate(input.toLowerCase(), /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/, 'email');
  };

  validations.password = function(input, method) {
    method = method || function(text) {
      return text.length > 5;
    };

    return validations.validate(input, method, 'password');
  };

  validations.passwordConfirm = function(input, confirm) {
    return validations.validate(input, confirm, 'password-confirm');
  };

  validations.passwordStrength = function(password, secureThreshold) {
    if(!password.length) return 0;

    // Sets default threshold to NIST (SP 800-63) recommended 80bit.
    secureThreshold = secureThreshold || 80;

    var entropy = (password.length * Math.log(alphaSize(password))) / Math.log(2);
    var bits = Math.round(entropy * 100) / 100;

    if(bits > secureThreshold) bits = secureThreshold;

    return ~~((bits / secureThreshold) * 100);
  };

  validations.maxLength = function(input, length) {
    length = length || 0;

    return validations.validate(input, function(text) {
      var type = typeof text;

      if(type === 'undefined' || text === null) {
        return false;
      }

      if(text instanceof Array || type === 'string') {
        return text.length <= length;
      }

      return false;
    }, 'maxLength');
  };

  validations.url = function(input) {
    return validations.validate(input, function(text) {
      return (/^(ftp|http|https):\/\/(\w+:{0,1}\w*@)?(\S+)(:[0-9]+)?(\/|\/([\w#!:.?+=&%@!\-\/]))?$/gi).test(text);
    }, 'url');
  };

  if (typeof exports !== 'undefined') {
    if (typeof module !== 'undefined' && module.exports) {
      exports = module.exports = validations;
    }
    exports.validations = validations;
  } else {
    root.validations = validations;
  }


}());
